zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt

```diff
@@ -28,38 +31,72 @@ import okio.IOException
 import okio.Path
 
 /**
- * Gets code from an HTTP server or a local cache,
- * and loads it into a zipline instance. This attempts
- * to load code as quickly as possible, and will
- * concurrently download and load code.
+ * Gets code from an HTTP server, or optional local cache
+ * or embedded filesystem, and loads it into a zipline instance.
+ * Loader attempts to load code as quickly as possible with
+ * concurrent network downloads and code loading.
  */
-class ZiplineLoader(
+class ZiplineLoader private constructor(
   private val dispatcher: CoroutineDispatcher,
   private val httpClient: ZiplineHttpClient,
-  private val embeddedDir: Path,
-  private val embeddedFileSystem: FileSystem,
-  cacheDbDriver: SqlDriver, // SqlDriver is already initialized to the platform and SQLite DB on disk
-  cacheDir: Path,
-  val cacheFileSystem: FileSystem,
-  cacheMaxSizeInBytes: Int = 100 * 1024 * 1024,
-  nowMs: () -> Long, // 100 MiB
+  private val fsCachingFetcher: FsCachingFetcher? = null,
+  private val fsEmbeddedFetcher: FsEmbeddedFetcher? = null,
 ) {
+  constructor(dispatcher: CoroutineDispatcher, httpClient: ZiplineHttpClient) : this(
+    dispatcher = dispatcher,
+    httpClient = httpClient,
+    fsCachingFetcher = null,
+    fsEmbeddedFetcher = null
+  )
+
+  fun withCache(
+    cacheDbDriver: SqlDriver, // SqlDriver is already initialized to the platform and SQLite DB on disk
+    cacheDir: Path,
+    cacheFileSystem: FileSystem,
+    cacheMaxSizeInBytes: Int = 100 * 1024 * 1024,
+    nowMs: () -> Long, // 100 MiB
+  ): ZiplineLoader = ZiplineLoader(
```

Comment:
I would prefer these with-ers move to be constructors on their respective fetchers. The loader ideally should not know anything about any subtypes of fetcher and it should only accept a single fetcher. Outside of the loader, you can compose fetchers together in any order and from any source that you want. Then that composite fetcher is the sole one supplied when creating a loader.

Reply:
I thought the overall API was simpler with fetchers being implementation details . . . . do callers want to assemble their own thing or get something more opinionated? I prefer opinionated here

Reply:
I'm not sure I'd call an API where you have to supply the embedded filesystem and path; the cache filesystem, path, database, size, and clock; and the http client as simple. You're creating rigidity in the implementation while also exposing the implementation in the API.

The implementation keeps changing and many of the tests have to change with it because so much of the implementation detail is plumbed out through the API.

I don't think what I wrote is meaningfully distant from the current with-ers API but it's definitely more flexible and definitely more testable. Each fetcher can be tested in isolation and they don't have to live in this library. A composite fetcher can be tested in isolation to ensure it correctly falls back. And the loader integration to the fetcher can be tested without concern for what the fetch does.